internal/command/program_resolver.go

```diff
@@ -0,0 +1,319 @@
+package command
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"slices"
+	"strings"
+
+	"mvdan.cc/sh/v3/syntax"
+)
+
+type ProgramResolverMode uint8
+
+const (
+	// unspecified is auto (default) which prompts for all unresolved
+	// subsequent runs will likely resolve via session
+	ProgramResolverModeAuto ProgramResolverMode = iota
+	// always prompt even if resolved
+	ProgramResolverModePrompt
+	// don't prompt whatsover even unresolved is resolved
+	ProgramResolverModeSkip
+)
+
+type ProgramResolverSource func() []string
+
+func ProgramResolverSourceFunc(env []string) ProgramResolverSource {
+	return func() []string {
+		return env
+	}
+}
+
+// ProgramResolver uses a list of ProgramResolverSource to resolve environment variables
+// found in a shell program. The result contains all found environment variables.
+// If the env is in any source, it is considered resolved. Otherwise, it is marked
+// as unresolved.
+type ProgramResolver struct {
+	mode           ProgramResolverMode
+	sources        []ProgramResolverSource
+	envCache       map[string]string
+	printer        *syntax.Printer
+	modifiedScript bool
+}
+
+func NewProgramResolver(mode ProgramResolverMode, sources ...ProgramResolverSource) *ProgramResolver {
+	return &ProgramResolver{
+		mode:     mode,
+		sources:  sources,
+		envCache: nil,
+		printer:  syntax.NewPrinter(),
+	}
+}
+
+type ProgramResolverPrompt uint8
+
+const (
+	ProgramResolverUnresolved ProgramResolverPrompt = iota
+	ProgramResolverResolved
+	ProgramResolverMessage
+	ProgramResolverPlaceholder
+)
+
+type ProgramResolverResult struct {
+	Prompt        ProgramResolverPrompt
```

Comment:
@sourishkrout this is hard to understand conceptually what `Prompt` really is. It seems like it's an indicator how the prompt in the VS Code should look like, based on the mode and whether an env is found in the sources. Is this correct?

If so, I would rename `ProgramResolverPrompt` to `ProgramResolverStatus` and change enum values to:
* `ProgramResolverStatusResolved` meaning that the env is resolved and no further action is needed.
* `ProgramResolverStatusUnresolvedWithMessage` meaning that the env is unresolved but there is a message to display.
* `ProgramResolverStatusUnresolvedWithPlaceholder` meaning that the env is unresolved but there is a placeholder/default value.

Reply:
Correct. I actually had it called `Status` before and renamed it to `Prompt` but can't remember the reason for it.

The renaming makes sense and is definitely more self-describing 👍 .